### PR TITLE
WIP Add Dockerfile to setup Docker image

### PR DESCRIPTION
**Pull request by madadam**
**DO NOT MERGE** yet, work in progress.

Pending stuff:

- [x] ~~decide whether to install safe_vault from APT repo, or from crates.io~~ Install from crates.io is the way to go
- [ ] add up-to-date seed node contacts to the crust config file
- [x] ~~figure out how to make service discovery work from outside the docker container (might not be necessary)~~ Containers running on the same network can already access local discovery. For container on different networks a possible solution would be to setup a multi-host networking (docker docs have info on this).
- [ ] Once new vault version is published, change the `cargo install` command to pull the latest published version, not master

 

---
This change is [Reviewable](https://reviewable.io/reviews/maidsafe/safe_vault/414)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @maqi

(maidsafe_highfive has picked a reviewer for you, use r? to override)

**Comment by dirvine** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[installer/docker/Dockerfile, line 22 \[r2\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFB5jo0CegvTsAshrm7:-KFB5jo0CegvTsAshrm8:642685005) ([raw file](https://github.com/maidsafe/safe_vault/blob/75e925755d2d67ecd3b21385ffcef57624331c35/installer/docker/Dockerfile#L22)):*
I think @madadam we do not need to do this. What we need is to install the vault binary only. Perhaps best to get that updated in the apt repo @fraser999 should be able to help. This image here should be perhaps an image to help folk code in rust on our vaults, but probably not what we want. So if this just installs the vault then we should be good. 
The next steps should be musl compiled packages - but that's another task, separate from this. This one just needs to install a vault and allow setting of wallet address and max_capacity in  the vault config file. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by madadam** [comment]
*[installer/docker/Dockerfile, line 22 \[r2\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFB5jo0CegvTsAshrm7:-KFETb_JX5lwlhLyibm9:133509872) ([raw file](https://github.com/maidsafe/safe_vault/blob/75e925755d2d67ecd3b21385ffcef57624331c35/installer/docker/Dockerfile#L22)):*
OK, after a bit of reading into this, I'm not convinced that installing safe-vault from the APT repo is the best way to go here. The vault in the apt is designed to run as a service in the background, and this is not the best way to run stuff inside docker containers. There are several issues with it:

- It would be tricky to detach the docker container (`docker run -d ...`), because the safe-vault process runs in the background, docker would think the root process terminated and would immediately terminated the container upon detaching it. See here: https://docs.docker.com/engine/reference/run/#detached-vs-foreground for more info about this.
- Accessing the log would be problematic. The safe_vault service logs somewhere into /var/log/... inside the container, and it's not easy to get this log from outside of the container

It's better if safe_vault is run as a normal program with its output directed to stdout/stderr and let docker handle detaching and log redirection.

So, the current `Dockerfile` builds the vault image like this:
1. Install everything necessary to build safe_vault from source (gcc, rustc, ...)
2. Pull the latest published safe_vault from crates.io (this can be changes to pull from the github repo instead, if desired)
3. Build and install it
4. Remove all the build dependencies, leaving only the bare minimum necessary to run safe_vault

The resulting image is very lightweight and can be started/stopped quickly (following the best practices: https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by dirvine** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[installer/docker/Dockerfile, line 22 \[r2\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFB5jo0CegvTsAshrm7:-KFEs-tTJQBU0IXy4byT:-267995255) ([raw file](https://github.com/maidsafe/safe_vault/blob/75e925755d2d67ecd3b21385ffcef57624331c35/installer/docker/Dockerfile#L22)):*
Neat job @madadam This does make a lot of sense and I like the approach of pulling from crates.io as well. Very cool, I am happy enough here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by dirvine** [comment]
Reviewed 1 of 3 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by dirvine** [comment]
Is it possible for people to set max capacity and perhaps wallet address via the  script @madadam  (of course with defaults available)  For now wallet address is not required but this would be a good test for when it is. If folks can run up many vaults with a wallet address they set then it will be very cool

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-:-KFEsMI6_7cirZAlLyz6:1356273585)*

**Comment by madadam** [comment]
I'll look into it, but from the top of my head, we have two options:
- pass those pieces of info as environment variables. We would then spin up the container like this:
`docker run -e "VAULT_WALLET_ADDRESS=xyz" -e "VAULT_MAX_CAPACITY=12345" msafe/vault`
- using volumes, so the user would be able to put some config file into the volume and vault would read from it (I have to read more about volumes first to know how exactly this would be done).

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-:-KFEtf8bWP3FQ0b9GRry:-1082561943)*

**Comment by dirvine** [comment]
I think option 1 looks good as it's perhaps easier to script for deployers etc. :+1: 

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-:-KFEwYUV5_q5tyYFsCJR:-1347766432)*

**Comment by madadam** [comment]
I agree. I need to write a tiny script that generates the vault config from those env variables, but that should be easy.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-:-KFEzNlKQzyO5kgFOIzR:-187057472)*

File: installer/docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM debian:jessie
+MAINTAINER Adam Cigánek <adam.ciganek@maidsafe.net>
```
Comment (dirvine) [inline installer/docker/Dockerfile]: please man MAidSafe Developers <dev@maidsfe.net>

File: installer/docker/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+## Safe Vault image
+
+Safe Vault is available as a docker image called **msafe/vault** on Docker Hub.
+Alternatively, it's possible to build the image using the Dockerfile found in `installer/docker`.
+
+To pull the latest Safe Vault image from Docker Hub:
+
+```
+docker pull msafe/vault
```
Comment (dirvine) [inline installer/docker/README.md]: This does not work atm ?

**Comment by dirvine** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[installer/docker/README.md, line 28 \[r5\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFiRHDaGDr16JaoZs92:-KFiRHDaGDr16JaoZs93:629566626) ([raw file](https://github.com/maidsafe/safe_vault/blob/790a8418e224f54ab1aa4d617055bfae17e4557d/installer/docker/README.md#L28)):*
would be nice to make this more random / automatic if possible?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by dirvine** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[installer/docker/README.md, line 31 \[r5\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFiRP2j_F35Ea6HC1TR:-KFiRP2j_F35Ea6HC1TS:-2085440690) ([raw file](https://github.com/maidsafe/safe_vault/blob/790a8418e224f54ab1aa4d617055bfae17e4557d/installer/docker/README.md#L31)):*
Aha :-) we should not really let folks change discovery port though, or at least make it easy. Perhaps saying change to discovery will not work with safe_vault networks on the SAFE network ? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by dirvine** [comment]
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[installer/docker/README.md, line 39 \[r5\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFiRhIDBDS223y7JLY4:-KFiRhIE2bYVNkIVqWOh:-1578054450) ([raw file](https://github.com/maidsafe/safe_vault/blob/790a8418e224f54ab1aa4d617055bfae17e4557d/installer/docker/README.md#L39)):*
Brilliant !!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by dirvine** [comment]
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[installer/docker/README.md, line 51 \[r5\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFiRmmvg8oQJ5imTt1M:-KFiRmmvg8oQJ5imTt1N:1419880617) ([raw file](https://github.com/maidsafe/safe_vault/blob/790a8418e224f54ab1aa4d617055bfae17e4557d/installer/docker/README.md#L51)):*
This is good and we will expand this with weekly updates as we progress. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by dirvine** [comment]
Reviewed 1 of 4 files at r4, 3 of 7 files at r5.
Review status: 3 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[installer/docker/safe_vault.vault.config, line 2 \[r5\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFiS0em4xJVUE3c8VGl:-KFiS0em4xJVUE3c8VGm:1968782653) ([raw file](https://github.com/maidsafe/safe_vault/blob/790a8418e224f54ab1aa4d617055bfae17e4557d/installer/docker/safe_vault.vault.config#L2)):*
should these default to `None` or `Null` @viv-rajkumar can confirm perhaps. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by madadam** [comment]
Review status: 3 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[installer/docker/README.md, line 14 \[r5\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFP_gbOdWxlPaFDHhmL-r5-14:-KFic8eEpLqYr0kmqcbX:525947721) ([raw file](https://github.com/maidsafe/safe_vault/blob/790a8418e224f54ab1aa4d617055bfae17e4557d/installer/docker/README.md#L14)):*
Nope, because the image is not yet published to Docker Hub. After we resolve all outstanding issues, we can publish it.

---

*[installer/docker/README.md, line 28 \[r5\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFiRHDaGDr16JaoZs92:-KFicYJOI0pcAScmX_Ba:-382688255) ([raw file](https://github.com/maidsafe/safe_vault/blob/790a8418e224f54ab1aa4d617055bfae17e4557d/installer/docker/README.md#L28)):*
Should be possible. We can use the `-P` flag to publish all exposed ports to random ports on the host machine. I'm going to verify that it would work and then change this instructions.

---

*[installer/docker/safe_vault.vault.config, line 2 \[r5\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFiS0em4xJVUE3c8VGl:-KFidByrusQg_2SYiBFA:-1097763396) ([raw file](https://github.com/maidsafe/safe_vault/blob/790a8418e224f54ab1aa4d617055bfae17e4557d/installer/docker/safe_vault.vault.config#L2)):*
Oops, this file shouldn't even be here. It's generated automatically inside the container, using the environment variables. I'll remove it.

Also, the crust config file need to be modified before publishing this. As a minimum, it needs at least one seed node contact in the `hard_coded_contacts` section.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by dirvine** [comment]
Reviewed 1 of 7 files at r5.
Review status: 4 of 6 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by Fraser999** [comment]
Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[installer/docker/safe_vault.crust.config, line 1 \[r7\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFoAMY13rdqwxngzwkb:-KFoAMY13rdqwxngzwkc:1298785618) ([raw file](https://github.com/maidsafe/safe_vault/blob/0a8a9f94a6471aa4f4eb8dfb3ca4ed90ed1ec506/installer/docker/safe_vault.crust.config#L1)):*
Can we use https://github.com/maidsafe/safe_vault/blob/master/installer/common/safe_vault.crust.config here instead?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by maqi** [comment]
*[a discussion](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFoAweO0kYiuuUq_Nxm:-KFoAwePhjGs-s4dFGi4:-464742917) (no related file):*
does all pending bullet points supposed to be all ticked off before merging this PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by madadam** [comment]
*[a discussion](https://reviewable.io:443/reviews/maidsafe/safe_vault/414#-KFoAweO0kYiuuUq_Nxm:-KFoN8O5Pj46M2JZSJCP:1550750685) (no related file):*
Yep, do not merge yet.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/414)*

**Comment by madadam** [comment]
Closing in favor of https://github.com/maidsafe/safe_vault/pull/451

**Pull request closed without merging**